tests/unittests/CoreGraphics.drawing/ImageHelpers.cpp

```diff
@@ -0,0 +1,114 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#include "ImageHelpers.h"
+
+#include <Starboard/SmartTypes.h>
+#include <ImageIO/ImageIO.h>
+
+#include <memory>
+
+#include <windows.h>
+
+CGImageRef _CGImageCreateFromPNGFile(CFStringRef filename) {
+    woc::unique_cf<CGDataProviderRef> dataProvider{ CGDataProviderCreateWithFilename(
+        CFStringGetCStringPtr(filename, kCFStringEncodingUTF8)) };
+    if (!dataProvider) {
+        return nullptr;
+    }
+
+    woc::unique_cf<CGImageRef> image{ CGImageCreateWithPNGDataProvider(dataProvider.get(), nullptr, FALSE, kCGRenderingIntentDefault) };
+    return image.release();
+}
+
+CFDataRef _CFDataCreatePNGFromCGImage(CGImageRef image) {
+    // Estimate the image data size to be as large as its raw pixel buffer.
+    // This will never be hit; but if it does, CFData will grow intelligently regardless.
+    size_t sizeEstimate = CGImageGetHeight(image) * CGImageGetBytesPerRow(image);
+    woc::unique_cf<CFMutableDataRef> imageData{ CFDataCreateMutable(nullptr, sizeEstimate) };
+    if (!imageData) {
+        return nullptr;
+    }
+
+    woc::unique_cf<CGImageDestinationRef> imageDest{ CGImageDestinationCreateWithData(imageData.get(), CFSTR("public.png"), 1, nullptr) };
+    if (!imageDest) {
+        return nullptr;
+    }
+
+    CGImageDestinationAddImage(imageDest.get(), image, nullptr);
+    CGImageDestinationFinalize(imageDest.get());
+
+    return imageData.release();
+}
+
+bool _WriteCFDataToFile(CFDataRef data, CFStringRef filename) {
+    std::unique_ptr<char[]> owningFilenamePtr;
+
+    char* rawFilename = const_cast<char*>(CFStringGetCStringPtr(filename, kCFStringEncodingUTF8));
+    size_t len = 0;
+
+    if (!rawFilename) {
+        CFRange filenameRange{ 0, CFStringGetLength(filename) };
+        CFIndex requiredBufferLength = 0;
+        CFStringGetBytes(filename, filenameRange, kCFStringEncodingUTF8, 0, FALSE, nullptr, 0, &requiredBufferLength);
+        owningFilenamePtr.reset(new char[requiredBufferLength]);
+        rawFilename = owningFilenamePtr.get();
+        CFStringGetBytes(
+            filename, filenameRange, kCFStringEncodingUTF8, 0, FALSE, (UInt8*)rawFilename, requiredBufferLength, &requiredBufferLength);
+        len = requiredBufferLength;
+    } else {
+        len = strlen(rawFilename);
+    }
+
+    woc::unique_cf<CFURLRef> url{ CFURLCreateFromFileSystemRepresentation(nullptr, (UInt8*)rawFilename, len, FALSE) };
+    if (!url) {
+        return false;
+    }
+
+    return CFURLWriteDataAndPropertiesToResource(url.get(), data, nullptr, nullptr);
+}
+
+CFDataRef _CFDataCreateFromCGImage(CGImageRef image) {
+    CGDataProviderRef provider{ CGImageGetDataProvider(image) };
```

Comment:
Probably not.

---
In reply to: [89019014](https://github.com/Microsoft/WinObjC/pull/1438#discussion_r89019014) [](ancestors = 89019014)